posthog/async_migrations/migrations/0002_events_sample_by.py

```diff
@@ -45,81 +65,80 @@ class Migration(AsyncMigrationDefinition):
-            resumable=True,
-            timeout_seconds=7 * 24 * 60 * 60,  # one week
-        ),
-        AsyncMigrationOperation(
-            fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
-            rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
-            resumable=True,
-        ),
-        AsyncMigrationOperation.simple_op(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            rollback=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-        ),
-        AsyncMigrationOperation.simple_op(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"""
-            INSERT INTO {TEMPORARY_TABLE_NAME}
-            SELECT * 
-            FROM {EVENTS_TABLE}
-            WHERE timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
-            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            resumable=True,
-            timeout_seconds=3 * 24 * 60 * 60,  # three days
-        ),
-        AsyncMigrationOperation.simple_op(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"""
-                RENAME TABLE
-                    {EVENTS_TABLE_NAME} to {BACKUP_TABLE_NAME},
-                    {TEMPORARY_TABLE_NAME} to {EVENTS_TABLE_NAME}
-                ON CLUSTER {CLICKHOUSE_CLUSTER}
-            """,
-            rollback=f"""
-                RENAME TABLE
-                    {EVENTS_TABLE_NAME} to {EVENTS_TABLE_NAME}_failed,
-                    {BACKUP_TABLE_NAME} to {EVENTS_TABLE_NAME}
-                ON CLUSTER {CLICKHOUSE_CLUSTER}
-            """,
-        ),
-        AsyncMigrationOperation.simple_op(
-            database=AnalyticsDBMS.CLICKHOUSE, sql=f"OPTIMIZE TABLE {EVENTS_TABLE_NAME} FINAL", resumable=True,
-        ),
-        AsyncMigrationOperation.simple_op(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            rollback=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-        ),
-        AsyncMigrationOperation(
-            fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
-            rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
-            resumable=True,
-        ),
-    ]
+    @cached_property
+    def operations(self):
+
+        create_table_op = [
+            AsyncMigrationOperation.simple_op(
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"""
+                CREATE TABLE IF NOT EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER} AS {EVENTS_TABLE_NAME}
+                ENGINE = ReplacingMergeTree(_timestamp)
+                PARTITION BY toYYYYMM(timestamp)
```

Comment:
might need to change this per now switching to weekly (https://posthog.slack.com/archives/C01MM7VT7MG/p1642496986079600) and make the migration required to everyone who doesn't have this partitioning?

Reply:
No need. That was just a slack discussion - let's not block in-progress PRs on it. If it happens, then separately from this.

Reply:
If we're doing this for cloud we should do this for self-hosted too. Or are we not planning to do this for cloud in the end in the near future?

Reply:
I haven't decided. Again - this conversation is not relevant for this PR, instead we should focus on getting it in. If we decide to do it we will just modify this as needed.